controllers/metal3.io/baremetalhost_controller_test.go

```diff
@@ -1854,3 +1861,66 @@ func TestUpdateRAID(t *testing.T) {
 		})
 	}
 }
+
+// ReconcileHelper is a test helper class to support verifying
+// specific reconciliation sequences step-by-step and for
+// improving test readability
+type ReconcileHelper struct {
+	t          *testing.T
+	host       *metal3v1alpha1.BareMetalHost
+	reconciler *BareMetalHostReconciler
+	request    reconcile.Request
+}
+
+func newReconcileHelper(t *testing.T, host *metal3v1alpha1.BareMetalHost) *ReconcileHelper {
+	return &ReconcileHelper{
+		t:          t,
+		host:       host,
+		reconciler: newTestReconciler(host),
+		request:    newRequest(host),
+	}
+}
+
+func (r *ReconcileHelper) verifyNextReconcile(expectedState metal3v1alpha1.ProvisioningState, expectedRequeue bool, expectedRequeueAfter int) {
+	res, err, host := tryReconcileSingleStep(r.reconciler, r.host, r.request)
+	assert.NoError(r.t, err)
+	assert.Equal(r.t, expectedRequeue, res.Requeue)
+	assert.Equal(r.t, time.Duration(expectedRequeueAfter), res.RequeueAfter)
+	assert.Equal(r.t, expectedState, host.Status.Provisioning.State)
+}
+
+func (r *ReconcileHelper) verifyNextReconcileOk(expectedState metal3v1alpha1.ProvisioningState) {
+	r.verifyNextReconcile(expectedState, true, 0)
+}
+
+func (r *ReconcileHelper) verifyNextReconcileStops(expectedState metal3v1alpha1.ProvisioningState) {
+	r.verifyNextReconcile(expectedState, false, 0)
+}
+
+func (r *ReconcileHelper) verifyNextReconcileError(expectedState metal3v1alpha1.ProvisioningState, expectedErrorMessage string) {
+	_, err, host := tryReconcileSingleStep(r.reconciler, r.host, r.request)
+	assert.EqualError(r.t, err, expectedErrorMessage)
+	assert.Equal(r.t, expectedState, host.Status.Provisioning.State)
+}
+
+func (r *ReconcileHelper) doDeleteHost() {
+	now := metav1.Now()
+	r.host.DeletionTimestamp = &now
+	r.host.Finalizers = []string{}
+	r.reconciler.Client.Update(context.Background(), r.host)
+}
+
+func TestInvalidBMHCanBeDeleted(t *testing.T) {
+	host := newDefaultHost(t)
+	host.Spec.BMC.Address = fmt.Sprintf("%s%s%s", "<", host.Spec.BMC.Address, ">")
+	r := newReconcileHelper(t, host)
+
+	r.verifyNextReconcileOk(metal3v1alpha1.StateNone)
```

Comment:
Can't this be done by

```
tryReconcile(t, r, h, func(host *metal3v1alpha1.BareMetalHost, result reconcile.Result) bool {
    return host.Status.Provisioning.State == metal3v1alpha1.StateNone
})
```

without factoring out `tryReconcileSingleStep()`?

Reply:
I'm looking at them as two different testing styles focusing on different elements. Here the sequence is completely unrolled to highlight each step, as I felt was important to follow the whole path (at least from the provisioning state point of view). Also this gives the opportunity to introduce in a simpler way "asynchronous" operations in the flow, like the Delete one. That's something that definitely could be achieved also with the `tryReconcile`, but still a couple of them would be needed.
I took also the opportunity to promote an (opinionated) scenario helper as an approach to push for test readability, given that it hides the complexity of the setup details and offers some verification/action methods where all the technical details have been stripped out.

Reply:
I'm not against it if it's generally useful. However, while probably fine in this instance, in general this is a brittle way to develop tests - the exact number of requeues gets hard-coded in the test, with the result that any change requires a rewrite of the tests. In most cases the thing we want to test is that the state converges to the expected one within a bounded number of requeues. So I'm not sure we want to encourage this style.
FWIW I believe there are other tests that do multiple tryReconcile()s with other operations intervening in the flow, so there is precedent for doing it that way.